Refactor to pluginManagement DSL

ethankhall opened the pull request:
This supersede #883 and #952.

This migrates the api to be more like:

```
pluginManagement {
  resolutionStrategy.eachPlugin { //not sure about these names yet, I just mimicked dependency resolution rules for now
    if (requested.id.startsWith('my.foo.'))
      target.useArtifact("com.mycompany:foo-plugins:$requested.version")
  }
  repositories {
    maven {
      url = "..."
    }
  }
}
```

ethankhall commented:
Hey @oehme @eriwen @alkemist @adammurdoch,

Here's my first pass at this. There are a few tests that are failing that I need to investigate why (related to the parsing of the settings.gradle file) and it needs more tests in general. But I wanted to get this to you so you could give feedback on the impl/dsl. Also there's a class that's poorly named, any suggestions on class name would be great thanks!

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/management/PluginManagementSpec.java:
```diff
@@ -0,0 +1,36 @@
+     * @param repositoriesAction spec to configure {@link RepositoryHandler}
+     * @since 3.4
+     */
+    void repositories(Action<? super RepositoryHandler> repositoriesAction);
```
This should be the `PluginRepositoryHandler` and `Settings.pluginRepositories {}` should delegate here.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/management/PluginResolutionSpec.java:
```diff
@@ -0,0 +1,15 @@
+ * to be loaded.
+ */
+@Incubating
+public interface PluginResolutionSpec {
```
This could be called `ConfigurablePluginRequest` and include things like `setVersion`.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/management/PluginResolutionSpec.java:
```diff
@@ -0,0 +1,15 @@
+@Incubating
+public interface PluginResolutionSpec {
+
+    void usePluginName(String name);
```
Can we leave this feature for a follow-up PR? I'd like to keep this one as small as we can.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/management/PluginResolveDetails.java:
```diff
@@ -0,0 +1,37 @@
+     *
+     * @since 3.4
+     */
+    void useTarget(Object notation);
```
How about a single method `changeTarget(Action<? super ConfigurablePluginRequest> action)`? The user would get a copy of the current target and could mutate it to his liking, e.g. setting the dependency notation, changing the plugin version etc.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/repository/PluginRepositoriesSpec.java:
```diff
@@ -48,5 +48,6 @@
      * @return The added repository.
      * @throws IllegalArgumentException if called more than once.
      */
+    @Deprecated
```
This should not be deprecated.

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/management/internal/DefaultPluginManagementPluginResolver.java:
```diff
@@ -0,0 +1,37 @@
+import org.gradle.plugin.use.resolve.internal.PluginResolutionResult;
+import org.gradle.plugin.use.resolve.internal.PluginResolveContext;
+
+public class DefaultPluginManagementPluginResolver implements PluginManagementPluginResolver {
```
Plugin management is not a new resolver. It is always executed on the plugin requests that were made by the user, potentially replacing them with different/enriched requests. Then these new requests are passed on to the existing resolvers. The existing resolvers can then make new decisions based on the richer request data. E.g. the Maven/Ivy resolver can look if the an explicit dependencyNotation was set and use that one instead of using their default logic.

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/management/internal/PluginResolutions.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @Nullable
+    public ResolvedPluginRequest resolveRequest(PluginRequest pluginRequest) {
```
This should never return null. If the user doesn't change the original request, then the original request is returned here.

oehme commented on subprojects/core/src/integTest/groovy/org/gradle/initialization/InitScriptIntegrationTest.groovy:
```diff
@@ -74,7 +77,59 @@ class InitScriptIntegrationTest extends AbstractIntegrationSpec {
-    private String initScript() {
+    def 'init script can specify plugin repositories'() {
+        executer.requireOwnGradleUserHomeDir()
+        new TestFile(executer.gradleUserHomeDir, "init.gradle") << "gradle.pluginRepositories { maven { it.url '${normaliseFileSeparators(mavenRepo.rootDir.absolutePath)}'} }"
```
This seems to be left over from a previous change.

ethankhall commented:
Doing the changes and I keep running into 
`Could not find method url() for arguments [http://repo.internal.net/m2] on object of type org.gradle.plugin.repository.internal.DefaultPluginRepositoriesSpec.` Do you guys have any idea? It seems like the closure coercion isn't working right. If I add a few methods that accept closure's that seems to fix the issue. So we need to fix that before a merge, but the tests that were written now work. (Update coming soon with the Closure's)

oehme commented:
Closure methods are added by DSL decoration. You need to use the `Instantiator` to create those objects instead of calling `new` directly.

oehme commented on subprojects/core/src/main/java/org/gradle/initialization/DefaultSettings.java:
```diff
@@ -275,4 +284,29 @@ public void buildCache(Action<? super BuildCacheConfiguration> action) {
+
+    @Override
+    public PluginRepositoriesSpec getPluginRepositoriesSpec() {
+        Instantiator instantiator = services.get(Instantiator.class);
```
I think we can get rid of this cruft now that we are in the `Settings` scope and use our normal service injection instead of creating a new instance every time. I.e. `PluginRepositoriesSpec` and `PluginManagementSpec` can live in the `SettingsScopedServices` and just be injected here, like the build cache configuration a few lines above.

ethankhall replied:
I tried that, but couldn't get a FileResolver that had a base dir so you could do relative paths. Do you know how to get a base dir to make this work? I tried StartParameters but rootDir was null.

oehme replied:
Okay, no problem. We can take a look when we merge and polish this.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/management/PluginManagementSpec.java:
```diff
@@ -0,0 +1,41 @@
+     */
+    void repositories(Action<? super PluginRepositoriesSpec> repositoriesAction);
+
+    void repositories(@DelegatesTo(value = PluginRepositoriesSpec.class, strategy = Closure.DELEGATE_FIRST) Closure closure);
```
Closure methods are added by the Instantiator, no need to write them by hand.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/management/PluginManagementSpec.java:
```diff
@@ -0,0 +1,41 @@
+
+    void repositories(@DelegatesTo(value = PluginRepositoriesSpec.class, strategy = Closure.DELEGATE_FIRST) Closure closure);
+
+    PluginResolutionStrategy getPluginResolutionStrategy();
```
Missing Action overload

oehme replied:
Should just be `resolutionStrategy`

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/management/PluginManagementSpec.java:
```diff
@@ -0,0 +1,41 @@
+     * @param repositoriesAction spec to configure {@link RepositoryHandler}
+     * @since 3.5
+     */
+    void repositories(Action<? super PluginRepositoriesSpec> repositoriesAction);
```
Missing getter

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequest.java:
```diff
@@ -17,26 +17,27 @@
     private final boolean apply;
     private final int lineNumber;
     private final String scriptDisplayName;
+    private final ConfiguredOptions configuredOptions = new ConfiguredOptions();
```
We should create a new, mutable plugin request and then make it immutable again once the configuration action is done. The plugin request should not have a mutable nested object.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/management/internal/InternalPluginResolutionStrategy.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface InternalPluginResolutionStrategy extends PluginResolutionStrategy {
+
+    void resolvePluginRequest(InternalPluginRequest pluginRequest);
```
This method needs to return a (potentially new) PluginRequest, not mutate the existing one.

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/management/internal/DefaultPluginResolveDetails.java:
```diff
@@ -0,0 +1,26 @@
+
+public class DefaultPluginResolveDetails implements PluginResolveDetails {
+
+    private final ConfigurablePluginRequest pluginRequest;
```
This needs to have two fields. The `requested`, immutable plugin request and the `target`, also immutable plugin request.

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/management/internal/DefaultPluginResolveDetails.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public void useTarget(Action<? super ConfigurablePluginRequest> action) {
+        action.execute(pluginRequest);
```
This method should create a new, mutable request, run the action and then make it immutable and assign it to the `target` field.

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/use/resolve/service/internal/ResolutionServiceResolver.java:
```diff
@@ -0,0 +1,82 @@
+import java.io.File;
+import java.util.Set;
+
+public class ResolutionServiceResolver {
```
There are quite a few changes to the plugin portal resolver which seem unrelated to this change. Could you try to remove these from the PR?

ethankhall replied:
Yeah will do!

oehme commented on subprojects/core/src/main/java/org/gradle/api/initialization/Settings.java:
```diff
@@ -204,4 +206,33 @@
+     * @throws IllegalStateException if a plugin has already been applied with the <code>plugins { }</code> block.
+     * @since ??
+     */
+    void pluginRepositories(Action<? super PluginRepositoriesSpec> pluginSettings);
```
Should not be here, we only want the new pluginManagement on the `Settings` DSL. The old `pluginRepositories` method should stay on the [InitialPassSettingsScript](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/initialization/InitialPassSettingsScript.java#L47), delegate to plugin management and be deprecated.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/management/ConfigurablePluginRequest.java:
```diff
@@ -0,0 +1,17 @@
+
+    void setVersion(String version);
+
+    void useTarget(String target);
```
Should probably take an Object just like dependency resolution rules do. Should be parsed into a `ModuleVersionSelector` immediately, so we don't leak potentially mutable stuff deep into the resolver infrastructure.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/use/PluginId.java:
```diff
@@ -0,0 +1,87 @@
+ * @since 3.4
+ */
+@Incubating
+public interface PluginId {
```
I don't think we need all these methods on the public API. At most we need `namespace`, `simpleName` and `qualifiedName`.

ethankhall replied:
I just extracted PluginId as an interface and left the impl as DefaultPluginId. Do we need to provide that abstraction layer?

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/use/internal/DefaultConfigurablePluginRequest.java:
```diff
@@ -0,0 +1,112 @@
+public class DefaultConfigurablePluginRequest implements InternalConfigurablePluginRequest {
+
+    private final InternalPluginRequest origonalRequest;
+    private boolean latch;
```
Not necessary. If we create one of these, then it means the user is about to configure something in all cases.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/use/internal/DefaultConfigurablePluginRequest.java:
```diff
@@ -0,0 +1,112 @@
+
+public class DefaultConfigurablePluginRequest implements InternalConfigurablePluginRequest {
+
+    private final InternalPluginRequest origonalRequest;
```
Not necessary. This should be initialized with the values from the original one and should have a method that converts it back to an immutable `DefaultPluginRequest`.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequest.java:
```diff
@@ -57,6 +57,17 @@ public String getVersion() {
+    }
+
+    @Override
+    public boolean isConfigured() {
```
Should not be necessary.

ethankhall replied:
The idea here was we could show if when a plugin fails to apply that it was enriched (so the user may know that something changed)

oehme replied:
I wouldn't do that, as it is kind of half-baked. If you really wanted tracing through resolution rules, you'd need to record every mutation. We don't even do that for normal resolution rules yet.

What we could do (but what I'd leave out of this PR since it is big enough) is passing *both* the requested and the target selector into the resolver chain, so they could do better reporting.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequest.java:
```diff
@@ -57,6 +57,17 @@ public String getVersion() {
 
+    @Nullable
+    @Override
+    public String getArtifact() {
```
Can be on the public API of `PluginRequest`. Should be `ModuleVersionSelector`, not `String`.

ethankhall replied:
It should be Object vs String but I don't think it will be resolved from what the user specified by then, unless you want me to call Dependencies.create when someone specifics an artifact.

oehme replied:
It should be a `ModuleVersionSelector`, just like in dependency resolution rules. A String is inconvenient, an Object would leak mutable stuff deeper into our resolver infrastructure.

You can use the `DependencyFactory` to convert from notation to `Dependency` and then create a `DefaultModuleVersionSelector` from that.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/use/internal/InternalConfigurablePluginRequest.java:
```diff
@@ -0,0 +1,25 @@
+
+import org.gradle.plugin.management.ConfigurablePluginRequest;
+
+public interface InternalConfigurablePluginRequest extends InternalPluginRequest, ConfigurablePluginRequest {
```
This interface should not be necessary.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/use/internal/InternalPluginRequest.java:
```diff
@@ -17,15 +17,21 @@
+import org.gradle.plugin.use.PluginId;
 
-public interface PluginRequest {
+public interface InternalPluginRequest extends PluginRequest {
```
This interface should not be necessary.

ethankhall replied:
Do we want to expose all the methods on this interface to user?

oehme replied:
There is only one that is not exposed and that is `getArtifact`. That should be exposed to rules, yes.

ethankhall replied:
I think it's necessary unless we want to expose a class to the consumers. If that's what you want then I can do it.

oehme replied:
I'm referring to `InternalPluginRequest`, which has no methods over the public `PluginRequest` interface.

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/management/internal/PluginResolutions.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    InternalPluginRequest resolveRequest(InternalPluginRequest pluginRequest) {
+        InternalConfigurablePluginRequest configurablePluginRequest = new DefaultConfigurablePluginRequest(pluginRequest);
```
We don't create a mutable request here. We pass in the immutable one and each rule returns an immutable one (possibly the same as we passed in). Only if the user actually mutates the target inside the rule, the rule will instantiate a mutable one and make it immutable once the user's action has run on it.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/repository/PluginRepositoriesSpec.java:
```diff
@@ -26,7 +27,7 @@
  */
 @Incubating
-public interface PluginRepositoriesSpec {
+public interface PluginRepositoriesSpec extends Configurable<PluginRepositoriesSpec> {
```
Extends not necessary. `Configurable` is internal API.

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/management/PluginResolveDetails.java:
```diff
@@ -0,0 +1,49 @@
+     * @return the requested plugin.
+     */
+    PluginRequest getRequestedPlugin();
+
```
`PluginRequest getTarget()` missing

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/use/PluginId.java:
```diff
@@ -0,0 +1,87 @@
+ * @since 3.5
+ */
+@Incubating
+public interface PluginId {
```
We should pare this down. I don't see much use for stuff like `maybeQualify` in user code. It should mirror what's available on `AppliedPlugin`, i.e. `id`, `namespace` and `name`.

oehme replied:
And to your question whether we need this: I think it is useful and we might actually change `AppliedPlugin` to have a `PluginId id()` method in the future and remove the duplicated methods over there.

ethankhall commented on subprojects/core/src/main/java/org/gradle/api/initialization/Settings.java:
```diff
@@ -204,4 +205,19 @@
+     *
+     * @return a {@link PluginManagementSpec}
+     */
+    PluginManagementSpec getPluginManagementSpec();
```
so no get? I thought groovy let you access methods as fields when you did something like this.

oehme replied:
Sorry I mean getPluginManagement. Drop the "Spec" :)

ethankhall commented:
I'll try to pull out things that aren't necessary here this weekend. I haven't addressed all of your feedback (mostly around naming) but I'll try to get it done tomorrow.

ethankhall commented:
I'm running into ```show onlyShortestPaths allResults
check absenceOfPackageCycles > 1 in org.gradle.*
  org.gradle.plugin.use.internal et al. contains 2 packages:
    org.gradle.plugin.use.internal
    org.gradle.plugin.management.internal``` when I try to make setArtifact parse the notation and give back a ModuleVersionSelector. Whats the recommended way to solve this?

oehme commented on subprojects/core/src/main/java/org/gradle/plugin/use/internal/InternalPluginRequest.java:
```diff
@@ -30,13 +30,13 @@
 
     String getDisplayName();
+
+    boolean isEnriched();
```
This method is not necessary. The resolver shouldn't care whether a request was modified by a rule or not. It just uses whatever the request says.

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/management/internal/PluginResolutions.java:
```diff
@@ -34,13 +32,17 @@ public void add(Action<? super PluginResolveDetails> rule) {
     InternalPluginRequest resolveRequest(InternalPluginRequest pluginRequest) {
-        InternalConfigurablePluginRequest configurablePluginRequest = new DefaultConfigurablePluginRequest(pluginRequest);
-        DefaultPluginResolveDetails details = new DefaultPluginResolveDetails(configurablePluginRequest);
+        if (resolutionDetails.isEmpty()) {
```
Unnecessary if-statement

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/use/resolve/internal/ArtifactRepositoryPluginResolver.java:
```diff
@@ -46,7 +46,7 @@ public void resolve(final InternalPluginRequest pluginRequest, PluginResolutionR
         }
 
-        if (!pluginRequest.isConfigured()) {
+        if (!pluginRequest.isEnriched()) {
```
The logic in the resolver is the same, no matter where the information in the request came from.

oehme commented on subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/repository/PluginManagementDslSpec.groovy:
```diff
@@ -0,0 +1,248 @@
+        fails 'help'
+
+        then:
+        failure.assertThatCause(containsString("Could not find method pluginManagement()"))
```
This shouldn't fail actually. `pluginManagement` is a normal method on `Settings`, so it will be available to imperative logic.

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/management/internal/PluginResolutions.java:
```diff
@@ -0,0 +1,48 @@
+
+class PluginResolutions {
+
+    private final Set<Action<? super PluginResolveDetails>> resolutionDetails = new LinkedHashSet<Action<? super PluginResolveDetails>>();
```
Probably should be called `resolutionRules`

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/management/internal/PluginResolutions.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+class PluginResolutions {
```
`PluginResolutionRules`?

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/management/internal/PluginResolutions.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    InternalPluginRequest resolveRequest(InternalPluginRequest pluginRequest) {
+        if (resolutionDetails.isEmpty()) {
```
unnecessary if-statement

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/use/resolve/internal/ArtifactRepositoryPluginResolver.java:
```diff
@@ -39,18 +39,22 @@ public ArtifactRepositoryPluginResolver(String name, DependencyResolutionService
-        if (pluginRequest.getVersion() == null) {
+    public void resolve(final InternalPluginRequest pluginRequest, PluginResolutionResult result) throws InvalidPluginRequestException {
+        String version = pluginRequest.getVersion();
+        if (version == null) {
```
Version is also optional if artifact is specified explicitly

oehme commented:
The package cycle seems unrelated to the `ModuleVersionSelector`. The error message says that you introduced a cycle between plugin.use and plugin.management.

> I also removed the isEnriched method, but that leads to the problem that we still have the issue that if a user wants to use a dynamic version they can't because we don't have a way to know that useVersion was called.

It doesn't matter who set the dynamic version. If a resolver doesn't support dynamic versions, there should be no way around that. We should also check this if the `artifact` is specified explicitly. I.e. if that `ModuleVersionSelector` has a dynamic/changing version, we should fail as well.

oehme commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/use/resolve/internal/ArtifactRepositoryPluginResolver.java:
```diff
@@ -39,18 +39,22 @@ public ArtifactRepositoryPluginResolver(String name, DependencyResolutionService
-            result.notFound(name, "dynamic plugin versions are not supported");
-            return;
+
+        if (null == pluginRequest.getArtifact()) {
```
The two checks below should always be done. Either on the plugin version or on the version on the `ModuleVersionSelector`. We don't support dynamic/changing versions in either case here.

ethankhall replied:
If a rule picks a dymanic/changing version why can't we support it? My use case is for local development I don't want to build final versions for testing so I'd want to use a snapshot.

oehme commented:
Thanks again for all your hard work @ethankhall. #1507 has been merged into `release` and will be part of Gradle 3.5
